dio deadlock in CI

from https://ci.vmware.run/vmware/vic/6065

```
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420076b00] read from closed multi-reader, returning EOF" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420076b00] read \"\" from 0 readers (err: &errors.errorString{s:\"EOF\"})" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] created multireader" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] read \"verify base multiwriter function\" from 2 readers (err: <nil>)" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removing reader - currently 2 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removed reader - now 1 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] Reader 0 [0xc42007a070]" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removing reader due to EOF" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removing reader - currently 1 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] removed reader - now 0 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] All of the readers returned EOF (1)" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] read \"\" from 1 readers (err: <nil>)" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] Going into sleep with 0 readers" 
time="2016-10-19T20:25:39Z" level=debug msg="[0xc420077000] read \"verify dynamic remove\" from 1 readers (err: <nil>)" 
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42006e0c0, 0x528b52, 0xe, 0x537990, 0xc42003bc01)
    /usr/local/go/src/testing/testing.go:647 +0x316
testing.RunTests.func1(0xc42006e0c0)
    /usr/local/go/src/testing/testing.go:793 +0x6d
testing.tRunner(0xc42006e0c0, 0xc42003bd88)
    /usr/local/go/src/testing/testing.go:610 +0x81
testing.RunTests(0x5379c0, 0x5b9b40, 0x8, 0x8, 0x40edae)
    /usr/local/go/src/testing/testing.go:799 +0x2f5
testing.(*M).Run(0xc42003bee8, 0x52dd80)
    /usr/local/go/src/testing/testing.go:743 +0x85
main.main()
    github.com/vmware/vic/pkg/dio/_test/_testmain.go:118 +0x1bd

goroutine 30 [semacquire]:
sync.runtime_notifyListWait(0xc420077050, 0xc400000000)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420077040)
    /usr/local/go/src/sync/cond.go:57 +0x80
github.com/vmware/vic/pkg/dio.(*multiReader).Read(0xc420077000, 0xc420097400, 0x15, 0x200, 0x0, 0x0, 0x0)
    github.com/vmware/vic/pkg/dio/_test/_obj_test/reader.go:76 +0x279
io.(*LimitedReader).Read(0xc42007cde0, 0xc420097400, 0x200, 0x200, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:436 +0x6c
bytes.(*Buffer).ReadFrom(0xc4200a4230, 0x5a7480, 0xc42007cde0, 0xc420089cb0, 0x1, 0x18)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200a4230, 0x5a7480, 0xc42007cde0, 0x0, 0x0, 0x0, 0x50c040, 0x7fa8f6d74001, 0xc42007cde0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200a4230, 0x5a7480, 0xc42007cde0, 0x507c80, 0x50b200, 0xc420089d00)
    /usr/local/go/src/io/io.go:360 +0x68
io.CopyN(0x5a7340, 0xc4200a4230, 0x7fa8f6d74078, 0xc420077000, 0x15, 0xc420077000, 0xc4200a4230, 0xc420078cc0)
    /usr/local/go/src/io/io.go:336 +0x8c
github.com/vmware/vic/pkg/dio.TestReadRemove(0xc420078cc0)
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:359 +0x8c8
testing.tRunner(0xc420078cc0, 0x537990)
    /usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
    /usr/local/go/src/testing/testing.go:646 +0x2ec

goroutine 17 [semacquire]:
sync.runtime_notifyListWait(0xc420078040, 0x0)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078030)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078000, 0xc420098020, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a000, 0xc420098020, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc4200820e0, 0x5a74c0, 0xc42007a000, 0xc420025ef8, 0xc420078001, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200820e0, 0x5a74c0, 0xc42007a000, 0x0, 0x0, 0x0, 0xc42007a008, 0x5a7500, 0xc42007a018)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200820e0, 0x5a74c0, 0xc42007a000, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestMultiWrite
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:43 +0x322

goroutine 18 [semacquire]:
sync.runtime_notifyListWait(0xc420078100, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc4200780f0)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc4200780c0, 0xc420098620, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a010, 0xc420098620, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082150, 0x5a74c0, 0xc42007a010, 0xc4200206f8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082150, 0x5a74c0, 0xc42007a010, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082150, 0x5a74c0, 0xc42007a010, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestMultiWrite
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:44 +0x36c

goroutine 20 [semacquire]:
sync.runtime_notifyListWait(0xc420078340, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078330)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078300, 0xc420098c32, 0x5ce, 0x5ce, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a020, 0xc420098c32, 0x5ce, 0x5ce, 0x12, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc4200823f0, 0x5a74c0, 0xc42007a020, 0xc420020ef8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc4200823f0, 0x5a74c0, 0xc42007a020, 0x0, 0x0, 0x0, 0x485f52, 0x5807d6c3, 0xc41ef17ce3)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc4200823f0, 0x5a74c0, 0xc42007a020, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteAdd
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:76 +0x2fa

goroutine 21 [semacquire]:
sync.runtime_notifyListWait(0xc420078400, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc4200783f0)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc4200783c0, 0xc420099212, 0x5ee, 0x5ee, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a030, 0xc420099212, 0x5ee, 0x5ee, 0x12, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082460, 0x5a74c0, 0xc42007a030, 0xc4200216f8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082460, 0x5a74c0, 0xc42007a030, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082460, 0x5a74c0, 0xc42007a030, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteAdd
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:77 +0x344

goroutine 23 [semacquire]:
sync.runtime_notifyListWait(0xc420078580, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078570)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078540, 0xc420099835, 0x5cb, 0x5cb, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a040, 0xc420099835, 0x5cb, 0x5cb, 0x15, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082a10, 0x5a74c0, 0xc42007a040, 0xc420021ef8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082a10, 0x5a74c0, 0xc42007a040, 0x0, 0x0, 0x0, 0xc42007a048, 0x5a7500, 0xc42007a058)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082a10, 0x5a74c0, 0xc42007a040, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteRemove
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:131 +0x322

goroutine 24 [semacquire]:
sync.runtime_notifyListWait(0xc420078640, 0x1)
    /usr/local/go/src/runtime/sema.go:267 +0x122
sync.(*Cond).Wait(0xc420078630)
    /usr/local/go/src/sync/cond.go:57 +0x80
io.(*pipe).read(0xc420078600, 0xc4200a0020, 0x5e0, 0x5e0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:47 +0x102
io.(*PipeReader).Read(0xc42007a050, 0xc4200a0020, 0x5e0, 0x5e0, 0x20, 0x0, 0x0)
    /usr/local/go/src/io/pipe.go:129 +0x4c
bytes.(*Buffer).ReadFrom(0xc420082a80, 0x5a74c0, 0xc42007a050, 0xc4200226f8, 0x1, 0x0)
    /usr/local/go/src/bytes/buffer.go:176 +0x155
io.copyBuffer(0x5a7340, 0xc420082a80, 0x5a74c0, 0xc42007a050, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:384 +0x323
io.Copy(0x5a7340, 0xc420082a80, 0x5a74c0, 0xc42007a050, 0x0, 0x0, 0x0)
    /usr/local/go/src/io/io.go:360 +0x68
created by github.com/vmware/vic/pkg/dio.TestWriteRemove
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:132 +0x36c

goroutine 34 [chan receive]:
github.com/vmware/vic/pkg/dio.TestReadRemove.func4(0xc420080f40, 0xc420090960, 0x5a8500, 0xc420077000, 0xc4200a4230, 0xc420078cc0)
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:349 +0x78
created by github.com/vmware/vic/pkg/dio.TestReadRemove
    /drone/src/github.com/vmware/vic/pkg/dio/multi_test.go:356 +0x852
FAIL    github.com/vmware/vic/pkg/dio   0.009s
make: *** [test-job-github.com/vmware/vic/pkg] Error 1
make: *** Waiting for unfinished jobs....
Makefile:242: recipe for target 'test-job-github.com/vmware/vic/pkg' failed
```

--- a/pkg/dio/multi_test.go
+++ b/pkg/dio/multi_test.go
@@ -1,4 +1,4 @@
-// Copyright 2016 VMware, Inc. All Rights Reserved.
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/pkg/dio/multi_test.go
+++ b/pkg/dio/multi_test.go
@@ -12,365 +12,278 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package io adds dynamic behaviour to the standard io package mutliX types
 package dio
 
 import (
 	"bytes"
 	"io"
-	"sync"
+	"strings"
 	"testing"
-
-	log "github.com/Sirupsen/logrus"
 )
 
-func init() {
-	// enable verbose logging during tests
-	log.SetLevel(log.DebugLevel)
-	verbose = true
+const (
+	count = 32
+
+	base    = "base functionality"
+	dynamic = "dynamic add/remove functionality"
+)
+
+func write(t *testing.T, mwriter DynamicMultiWriter, p []byte) {
+	n, err := mwriter.Write(p)
+	if err != nil {
+		t.Errorf("write: %v", err)
+	}
+	if n != len(p) {
+		t.Errorf("short write: %d != %d", n, len(p))
+	}
 }
 
+func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
+	total := 0
+
+	var buf = make([]byte, 32*1024)
+	for {
+		n, err := mreader.Read(buf)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			t.Errorf("read: %v", err)
+		}
+
+		total += n
+		if total >= limit {
+			break
+		}
+	}
+
+	return buf[:limit]
+}
+
+func control(t *testing.T, buffers []*bytes.Buffer, p []byte) {
+	for i := range buffers {
+		buffer := buffers[i]
+		if buffer.String() != string(p) {
+			t.Errorf("Expected: %q, actual: %q", string(p), buffer.String())
+		}
+	}
+}
+
+func equal(t *testing.T, mreader DynamicMultiReader, str string, count int) {
+	expected := strings.Repeat(str, count)
+	// read from multi reader
+	buffer := string(read(t, mreader, len(expected)))
+	// compare the data
+	if buffer != expected {
+		t.Errorf("Expected: %q, actual: %q", expected, buffer)
+	}
+}
+
+// TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
 func TestMultiWrite(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	// create & initialize writers and buffers
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+		reader, writer := io.Pipe()
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+		writers = append(writers, writer)
+		buffers = append(buffers, &buffer)
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
 }
 
+// TestMultiWrite creates bunch of multi writers and writes to them, then adds more and writes again
 func TestWriteAdd(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
 
-	mwriter := MultiWriter(pipeAW)
+	var writersAdded []io.Writer
+	var buffersAdded []*bytes.Buffer
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+	// create & initialize writers and buffers into two categories
+	// *Added ones will be added to multi writer later
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+		reader, writer := io.Pipe()
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+		// add the writer & buffer to skip list if i divisible by three
+		if i%3 == 0 {
+			writersAdded = append(writersAdded, writer)
+			buffersAdded = append(buffersAdded, &buffer)
+		} else {
+			writers = append(writers, writer)
+			buffers = append(buffers, &buffer)
+		}
+
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data - shouldn't be present
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != "" {
-		t.Errorf("B: expected: %s, actual: %s", "", bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
 
-	// Add writer to existing MultiWriter
-	mwriter.Add(pipeBW)
+	// add skipped writers to the writer
+	mwriter.Add(writersAdded...)
 
-	data2 := "verify dynamic add"
-	_, err = mwriter.Write([]byte(data2))
-	if err != nil {
-		t.Error(err)
-		return
-	}
-
-	// compare the data - shouldn't be present
-	if bufA.String() != data+data2 {
-		t.Errorf("A: expected: %s, actual: %s", data+data2, bufA.String())
-		return
-	}
-
-	if bufB.String() != data2 {
-		t.Errorf("B: expected: %s, actual: %s", data2, bufB.String())
-		return
-	}
-
+	write(t, mwriter, []byte(dynamic))
+	control(t, buffers, []byte(base+dynamic))
+	control(t, buffersAdded, []byte(dynamic))
 }
 
+// TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
 func TestWriteRemove(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	var buffersLeft []*bytes.Buffer
+	var buffersRemoved []*bytes.Buffer
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+	// create & initialize writers and buffers into two categories
+	// *Removed ones will be filtered out from multi writer later
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+		reader, writer := io.Pipe()
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+		writers = append(writers, writer)
+		buffers = append(buffers, &buffer)
+
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
+
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
+
+	// add the writer & buffer to skip list if i divisible by three
+	for i := 0; i < count; i++ {
+		if i%3 == 0 {
+			mwriter.Remove(writers[i])
+
+			buffersRemoved = append(buffersRemoved, buffers[i])
+		} else {
+			buffersLeft = append(buffersLeft, buffers[i])
+		}
 	}
 
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-		return
-	}
-
-	// Remove the writer from the MultiWriter
-	mwriter.Remove(pipeBW)
-
-	data2 := "verify dynamic remove"
-	_, err = mwriter.Write([]byte(data2))
-	if err != nil {
-		t.Error(err)
-		return
-	}
-
-	// compare the data - shouldn't be present
-	if bufA.String() != data+data2 {
-		t.Errorf("A: expected: %s, actual: %s", data+data2, bufA.String())
-		return
-	}
-
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(dynamic))
+	control(t, buffersLeft, []byte(base+dynamic))
+	control(t, buffersRemoved, []byte(base))
 }
 
-func TestWriteConcurrentRemove(t *testing.T) {
-	t.Skip("not sure how to test concurrency in this case")
-}
-
+// TestMultiRead creates multi readers and reads from them
 func TestMultiRead(t *testing.T) {
-	var wg sync.WaitGroup
+	var readers []io.Reader
 
-	dataA := "verify base multireader functionA"
-	dataB := "verify base multireader functionB"
+	// create & initialize writers and buffers
+	for i := 0; i < count; i++ {
+		reader, writer := io.Pipe()
 
-	readerA := bytes.NewReader([]byte(dataA))
-	readerB := bytes.NewReader([]byte(dataB))
+		readers = append(readers, reader)
 
-	mreader := MultiReader(readerA, readerB)
-
-	var buf bytes.Buffer
-
-	// do the read
-	_, err := io.CopyN(&buf, mreader, int64(len(dataA)+len(dataB)))
-	if err != nil {
-		t.Error(err)
+		// set up a  goroutine so we don't block reads
+		go io.Copy(writer, bytes.NewReader([]byte(base)))
 	}
 
-	// compare the data
-	if buf.String() != dataA+dataB {
-		t.Errorf("A: expected: %s, actual: %s", dataA+dataB, buf.String())
-		return
-	}
+	// create the multi writer
+	mreader := MultiReader(readers...)
 
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		_, err := mreader.Read(buf.Bytes())
-		if err != io.EOF {
-			t.Error(err)
-		}
-	}()
-
-	mreader.Close()
-	wg.Wait()
+	equal(t, mreader, base, count)
 }
 
+// TestMultiRead creates multi readers and reads from them, then adds mores and reads again
 func TestReadAdd(t *testing.T) {
-	var wg sync.WaitGroup
+	var readers []io.Reader
+	var pipereaders []io.PipeReader
 
-	done := make(chan struct{}, 1)
-	defer close(done)
+	var readersAdded []io.Reader
 
-	dataA := "verify base multireader functionA"
-	dataB := "verify base multireader functionB"
+	skipped := 0
+	// create & initialize writers and buffers
+	for i := 0; i < count; i++ {
+		reader, writer := io.Pipe()
 
-	readerA := bytes.NewReader([]byte(dataA))
-	readerB := bytes.NewReader([]byte(dataB))
+		if i%3 == 0 {
+			skipped++
 
-	mreader := MultiReader(readerA)
+			readersAdded = append(readersAdded, reader)
+			// set up a  goroutine so we don't block reads
+			go io.Copy(writer, bytes.NewReader([]byte(dynamic)))
+		} else {
+			pipereaders = append(pipereaders, *reader)
+			readers = append(readers, reader)
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
-
-	// do the read - bytes.NewReader does not return data and EOF
-	// from the same call, so this should have err==nil
-	_, err := io.CopyN(&bufA, mreader, int64(len(dataA)))
-	if err != nil {
-		t.Error(err)
-	}
-
-	// compare the data
-	if bufA.String() != dataA {
-		t.Errorf("A: expected: %s, actual: %s", dataA, bufA.String())
-		return
-	}
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-
-		<-done
-
-		_, err := mreader.Read(bufA.Bytes())
-		if err != io.EOF {
-			t.Error(err)
+			// set up a  goroutine so we don't block reads
+			go io.Copy(writer, bytes.NewReader([]byte(base)))
 		}
-	}()
-
-	// Add reader to existing MultiReader
-	// this should furnish new data to the copy without further action being
-	// taken
-	mreader.Add(readerB)
-
-	// do the read - we expect mreader to now switch to the new source, which
-	// has the standard bytes.NewReader behaviour
-	_, err = io.CopyN(&bufB, mreader, int64(len(dataB)))
-	if err != nil {
-		t.Error(err)
-	}
-	done <- struct{}{}
-
-	// compare the data
-	if bufB.String() != dataB {
-		t.Errorf("A: expected: %s, actual: %s", dataB, bufB.String())
-		return
 	}
 
-	mreader.Close()
-	wg.Wait()
+	// create the multi writer
+	mreader := MultiReader(readers...)
+
+	equal(t, mreader, base, count-skipped)
+
+	// add the rest of the readers
+	mreader.Add(readersAdded...)
+
+	// close the initial set otherwise they will block
+	for i := range pipereaders {
+		pipereaders[i].Close()
+	}
+
+	equal(t, mreader, dynamic, skipped)
 }
 
+// TestReadRemove creates multi readers and reads from them, then removes some and reads again
 func TestReadRemove(t *testing.T) {
-	var wg sync.WaitGroup
+	var readers []io.Reader
+	var writers []io.Writer
 
-	done := make(chan struct{}, 1)
-	defer close(done)
+	// create & initialize writers and buffers
+	for i := 0; i < count; i++ {
+		reader, writer := io.Pipe()
 
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+		readers = append(readers, reader)
+		writers = append(writers, writer)
 
-	mreader := MultiReader(pipeAR, pipeBR)
-
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
-
-	// send the test string
-	data := "verify base multiwriter function"
-
-	go func() {
-		defer pipeAW.Close()
-		_, err := pipeAW.Write([]byte(data))
-		if err != nil {
-			t.Error(err)
-		}
-	}()
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-
-		<-done
-
-		_, err := mreader.Read(bufA.Bytes())
-		if err != nil && err != io.EOF && err != io.ErrClosedPipe {
-			t.Error(err)
-		}
-	}()
-
-	// do the read
-	_, err := io.CopyN(&bufA, mreader, int64(len(data)))
-	if err != nil {
-		t.Error(err)
+		// set up a goroutine so we don't block reads
+		go io.Copy(writer, bytes.NewReader([]byte(base)))
 	}
 
-	done <- struct{}{}
+	// create the multi writer
+	mreader := MultiReader(readers...)
 
-	// compare the data
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-	}
+	equal(t, mreader, base, count)
 
-	// Remove the writer from existing MultiWriter
-	mreader.Remove(pipeAR)
+	removed := 0
+	for i := 0; i < count; i++ {
+		if i%3 == 0 {
+			removed++
 
-	data = "verify dynamic remove"
-
-	go func() {
-		defer pipeBW.Close()
-		_, err = pipeBW.Write([]byte(data))
-		if err != nil {
-			t.Error(err)
+			mreader.Remove(readers[i])
+		} else {
+			// set up another goroutine so we don't block reads
+			go io.Copy(writers[i], bytes.NewReader([]byte(dynamic)))
 		}
-	}()
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-
-		<-done
-
-		_, err := mreader.Read(bufB.Bytes())
-		if err != nil && err != io.EOF && err != io.ErrClosedPipe {
-			t.Error(err)
-		}
-
-	}()
-
-	// do the read
-	_, err = io.CopyN(&bufB, mreader, int64(len(data)))
-	if err != nil {
-		t.Error(err)
 	}
-	done <- struct{}{}
-
-	// compare the data - shouldn't be present
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-	}
-
-	mreader.Close()
-	wg.Wait()
-}
-
-func TestReadConcurrentRemove(t *testing.T) {
-	t.Skip("not sure how to test concurrency in this case")
+	equal(t, mreader, dynamic, count-removed)
 }

--- a/pkg/dio/multi_test.go
+++ b/pkg/dio/multi_test.go
@@ -18,7 +18,10 @@ import (
 	"bytes"
 	"io"
 	"strings"
+	"sync"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
 )
 
 const (
@@ -27,6 +30,41 @@ const (
 	base    = "base functionality"
 	dynamic = "dynamic add/remove functionality"
 )
+
+type filterf func(idx int) bool
+
+func FilterBuffers(in []*bytes.Buffer, fn filterf) []*bytes.Buffer {
+	var out []*bytes.Buffer
+
+	for i := 0; i < len(in); i++ {
+		if fn(i) {
+			out = append(out, in[i])
+		}
+	}
+	return out
+}
+
+func FilterWriters(in []io.Writer, fn filterf) []io.Writer {
+	var out []io.Writer
+
+	for i := 0; i < len(in); i++ {
+		if fn(i) {
+			out = append(out, in[i])
+		}
+	}
+	return out
+}
+
+func FilterReaders(in []io.Reader, fn filterf) []io.Reader {
+	var out []io.Reader
+
+	for i := 0; i < len(in); i++ {
+		if fn(i) {
+			out = append(out, in[i])
+		}
+	}
+	return out
+}
 
 func write(t *testing.T, mwriter DynamicMultiWriter, p []byte) {
 	n, err := mwriter.Write(p)
@@ -43,7 +81,7 @@ func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
 
 	var buf = make([]byte, 32*1024)
 	for {
-		n, err := mreader.Read(buf)
+		n, err := mreader.Read(buf[total:limit])
 		if err == io.EOF {
 			break
 		}
@@ -60,27 +98,16 @@ func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
 	return buf[:limit]
 }
 
-func control(t *testing.T, buffers []*bytes.Buffer, p []byte) {
+func each(t *testing.T, buffers []*bytes.Buffer, s string) {
 	for i := range buffers {
-		buffer := buffers[i]
-		if buffer.String() != string(p) {
-			t.Errorf("Expected: %q, actual: %q", string(p), buffer.String())
-		}
-	}
-}
-
-func equal(t *testing.T, mreader DynamicMultiReader, str string, count int) {
-	expected := strings.Repeat(str, count)
-	// read from multi reader
-	buffer := string(read(t, mreader, len(expected)))
-	// compare the data
-	if buffer != expected {
-		t.Errorf("Expected: %q, actual: %q", expected, buffer)
+		assert.Equal(t, buffers[i].String(), s)
 	}
 }
 
 // TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
 func TestMultiWrite(t *testing.T) {
+	var wg sync.WaitGroup
+
 	var writers []io.Writer
 	var buffers []*bytes.Buffer
 
@@ -93,24 +120,31 @@ func TestMultiWrite(t *testing.T) {
 		writers = append(writers, writer)
 		buffers = append(buffers, &buffer)
 
-		// set up a goroutine so we don't block writes
-		go io.Copy(&buffer, reader)
+		wg.Add(1)
+		go func() {
+			// set up a goroutine so we don't block writes
+			io.CopyN(&buffer, reader, int64(len(base)))
+
+			wg.Done()
+		}()
 	}
 
 	// create the multi writer
 	mwriter := MultiWriter(writers...)
 
 	write(t, mwriter, []byte(base))
-	control(t, buffers, []byte(base))
+	wg.Wait()
+
+	each(t, buffers, base)
 }
 
 // TestMultiWrite creates bunch of multi writers and writes to them, then adds more and writes again
 func TestWriteAdd(t *testing.T) {
+	var wg sync.WaitGroup
+	var wgAdded sync.WaitGroup
+
 	var writers []io.Writer
 	var buffers []*bytes.Buffer
-
-	var writersAdded []io.Writer
-	var buffersAdded []*bytes.Buffer
 
 	// create & initialize writers and buffers into two categories
 	// *Added ones will be added to multi writer later
@@ -119,40 +153,56 @@ func TestWriteAdd(t *testing.T) {
 
 		reader, writer := io.Pipe()
 
-		// add the writer & buffer to skip list if i divisible by three
-		if i%3 == 0 {
-			writersAdded = append(writersAdded, writer)
-			buffersAdded = append(buffersAdded, &buffer)
-		} else {
-			writers = append(writers, writer)
-			buffers = append(buffers, &buffer)
-		}
+		writers = append(writers, writer)
+		buffers = append(buffers, &buffer)
 
-		// set up a goroutine so we don't block writes
-		go io.Copy(&buffer, reader)
+		if i%3 != 0 {
+			wg.Add(1)
+		}
+		wgAdded.Add(1)
+		go func(i int) {
+			if i%3 != 0 {
+				io.CopyN(&buffer, reader, int64(len(base)))
+				wg.Done()
+			}
+			io.CopyN(&buffer, reader, int64(len(dynamic)))
+
+			wgAdded.Done()
+		}(i)
 	}
 
+	writersAdded := FilterWriters(writers, func(i int) bool { return i%3 == 0 })
+	writersLeft := FilterWriters(writers, func(i int) bool { return i%3 != 0 })
+
+	buffersAdded := FilterBuffers(buffers, func(i int) bool { return i%3 == 0 })
+	buffersLeft := FilterBuffers(buffers, func(i int) bool { return i%3 != 0 })
+
 	// create the multi writer
-	mwriter := MultiWriter(writers...)
+	mwriter := MultiWriter(writersLeft...)
 
 	write(t, mwriter, []byte(base))
-	control(t, buffers, []byte(base))
+	wg.Wait()
+
+	each(t, buffersLeft, base)
 
 	// add skipped writers to the writer
 	mwriter.Add(writersAdded...)
 
 	write(t, mwriter, []byte(dynamic))
-	control(t, buffers, []byte(base+dynamic))
-	control(t, buffersAdded, []byte(dynamic))
+
+	wgAdded.Wait()
+
+	each(t, buffersLeft, base+dynamic)
+	each(t, buffersAdded, dynamic)
 }
 
 // TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
 func TestWriteRemove(t *testing.T) {
+	var wg sync.WaitGroup
+	var wgRemoved sync.WaitGroup
+
 	var writers []io.Writer
 	var buffers []*bytes.Buffer
-
-	var buffersLeft []*bytes.Buffer
-	var buffersRemoved []*bytes.Buffer
 
 	// create & initialize writers and buffers into two categories
 	// *Removed ones will be filtered out from multi writer later
@@ -165,33 +215,55 @@ func TestWriteRemove(t *testing.T) {
 		buffers = append(buffers, &buffer)
 
 		// set up a goroutine so we don't block writes
-		go io.Copy(&buffer, reader)
+		wg.Add(1)
+		wgRemoved.Add(1)
+		go func(i int) {
+			// set up a goroutine so we don't block writes
+			io.CopyN(&buffer, reader, int64(len(base)))
+
+			wg.Done()
+
+			if i%3 == 0 {
+				wgRemoved.Done()
+				return
+			}
+
+			io.CopyN(&buffer, reader, int64(len(dynamic)))
+
+			wgRemoved.Done()
+		}(i)
 	}
 
 	// create the multi writer
 	mwriter := MultiWriter(writers...)
 
 	write(t, mwriter, []byte(base))
-	control(t, buffers, []byte(base))
+	wg.Wait()
 
-	// add the writer & buffer to skip list if i divisible by three
+	each(t, buffers, base)
+
+	// remove the writers
 	for i := 0; i < count; i++ {
 		if i%3 == 0 {
 			mwriter.Remove(writers[i])
-
-			buffersRemoved = append(buffersRemoved, buffers[i])
-		} else {
-			buffersLeft = append(buffersLeft, buffers[i])
 		}
 	}
 
 	write(t, mwriter, []byte(dynamic))
-	control(t, buffersLeft, []byte(base+dynamic))
-	control(t, buffersRemoved, []byte(base))
+	wgRemoved.Wait()
+
+	buffersLeft := FilterBuffers(buffers, func(i int) bool { return i%3 != 0 })
+	buffersRemoved := FilterBuffers(buffers, func(i int) bool { return i%3 == 0 })
+
+	each(t, buffersLeft, base+dynamic)
+	each(t, buffersRemoved, base)
+
 }
 
 // TestMultiRead creates multi readers and reads from them
 func TestMultiRead(t *testing.T) {
+	var wg sync.WaitGroup
+
 	var readers []io.Reader
 
 	// create & initialize writers and buffers
@@ -200,47 +272,68 @@ func TestMultiRead(t *testing.T) {
 
 		readers = append(readers, reader)
 
-		// set up a  goroutine so we don't block reads
-		go io.Copy(writer, bytes.NewReader([]byte(base)))
+		wg.Add(1)
+		go func() {
+			// set up a  goroutine so we don't block reads
+			io.CopyN(writer, bytes.NewReader([]byte(base)), int64(len(base)))
+
+			wg.Done()
+		}()
 	}
 
 	// create the multi writer
 	mreader := MultiReader(readers...)
 
-	equal(t, mreader, base, count)
+	expected := strings.Repeat(base, count)
+	buffer := read(t, mreader, len(expected))
+
+	wg.Wait()
+
+	assert.Equal(t, expected, string(buffer))
 }
 
 // TestMultiRead creates multi readers and reads from them, then adds mores and reads again
 func TestReadAdd(t *testing.T) {
+	var wg sync.WaitGroup
+	var wgAdded sync.WaitGroup
+
 	var readers []io.Reader
 	var pipereaders []io.PipeReader
 
-	var readersAdded []io.Reader
-
-	skipped := 0
 	// create & initialize writers and buffers
 	for i := 0; i < count; i++ {
 		reader, writer := io.Pipe()
 
-		if i%3 == 0 {
-			skipped++
+		readers = append(readers, reader)
 
-			readersAdded = append(readersAdded, reader)
-			// set up a  goroutine so we don't block reads
-			go io.Copy(writer, bytes.NewReader([]byte(dynamic)))
-		} else {
+		if i%3 != 0 {
 			pipereaders = append(pipereaders, *reader)
-			readers = append(readers, reader)
 
-			// set up a  goroutine so we don't block reads
-			go io.Copy(writer, bytes.NewReader([]byte(base)))
+			wg.Add(1)
 		}
+		wgAdded.Add(1)
+		go func(i int) {
+			if i%3 != 0 {
+				io.CopyN(writer, bytes.NewReader([]byte(base)), int64(len(base)))
+				wg.Done()
+			}
+			io.CopyN(writer, bytes.NewReader([]byte(dynamic)), int64(len(dynamic)))
+
+			wgAdded.Done()
+		}(i)
 	}
 
-	// create the multi writer
-	mreader := MultiReader(readers...)
+	readersAdded := FilterReaders(readers, func(i int) bool { return i%3 == 0 })
+	readersLeft := FilterReaders(readers, func(i int) bool { return i%3 != 0 })
 
-	equal(t, mreader, base, count-skipped)
+	// create the multi writer
+	mreader := MultiReader(readersLeft...)
+
+	expected := strings.Repeat(base, len(readersLeft))
+	buffer := read(t, mreader, len(expected))
+	wg.Wait()
+
+	assert.Equal(t, expected, string(buffer))
 
 	// add the rest of the readers
 	mreader.Add(readersAdded...)
@@ -250,11 +343,20 @@ func TestReadAdd(t *testing.T) {
 		pipereaders[i].Close()
 	}
 
-	equal(t, mreader, dynamic, skipped)
+	expected = strings.Repeat(dynamic, len(readersAdded))
+	buffer = read(t, mreader, len(expected))
+
+	wgAdded.Wait()
+
+	assert.Equal(t, expected, string(buffer))
+
 }
 
 // TestReadRemove creates multi readers and reads from them, then removes some and reads again
 func TestReadRemove(t *testing.T) {
+	var wg sync.WaitGroup
+	var wgRemoved sync.WaitGroup
+
 	var readers []io.Reader
 	var writers []io.Writer
 
@@ -265,25 +367,48 @@ func TestReadRemove(t *testing.T) {
 		readers = append(readers, reader)
 		writers = append(writers, writer)
 
-		// set up a goroutine so we don't block reads
-		go io.Copy(writer, bytes.NewReader([]byte(base)))
+		// set up a goroutine so we don't block writes
+		wg.Add(1)
+		wgRemoved.Add(1)
+		go func(i int) {
+			// set up a goroutine so we don't block writes
+			io.CopyN(writer, bytes.NewReader([]byte(base)), int64(len(base)))
+
+			wg.Done()
+
+			if i%3 == 0 {
+				wgRemoved.Done()
+				return
+			}
+
+			io.CopyN(writer, bytes.NewReader([]byte(dynamic)), int64(len(dynamic)))
+
+			wgRemoved.Done()
+		}(i)
+
 	}
 
 	// create the multi writer
 	mreader := MultiReader(readers...)
 
-	equal(t, mreader, base, count)
+	expected := strings.Repeat(base, count)
+	buffer := read(t, mreader, len(expected))
+	wg.Wait()
 
-	removed := 0
-	for i := 0; i < count; i++ {
-		if i%3 == 0 {
-			removed++
+	assert.Equal(t, expected, string(buffer))
 
-			mreader.Remove(readers[i])
-		} else {
-			// set up another goroutine so we don't block reads
-			go io.Copy(writers[i], bytes.NewReader([]byte(dynamic)))
-		}
+	readersLeft := FilterReaders(readers, func(i int) bool { return i%3 != 0 })
+	readersRemoved := FilterReaders(readers, func(i int) bool { return i%3 == 0 })
+
+	for i := range readersRemoved {
+		mreader.Remove(readersRemoved[i])
 	}
-	equal(t, mreader, dynamic, count-removed)
+
+	expected = strings.Repeat(dynamic, len(readersLeft))
+	buffer = read(t, mreader, len(expected))
+
+	wgRemoved.Wait()
+
+	assert.Equal(t, expected, string(buffer))
+
 }

--- a/pkg/dio/multi_test.go
+++ b/pkg/dio/multi_test.go
@@ -132,6 +132,7 @@ func TestMultiWrite(t *testing.T) {
 	// create the multi writer
 	mwriter := MultiWriter(writers...)
 
+	// write and ensure io.Copy returns
 	write(t, mwriter, []byte(base))
 	wg.Wait()
 
@@ -180,6 +181,7 @@ func TestWriteAdd(t *testing.T) {
 	// create the multi writer
 	mwriter := MultiWriter(writersLeft...)
 
+	// write and ensure io.Copy returns
 	write(t, mwriter, []byte(base))
 	wg.Wait()
 
@@ -188,8 +190,8 @@ func TestWriteAdd(t *testing.T) {
 	// add skipped writers to the writer
 	mwriter.Add(writersAdded...)
 
+	// write and ensure io.Copy returns
 	write(t, mwriter, []byte(dynamic))
-
 	wgAdded.Wait()
 
 	each(t, buffersLeft, base+dynamic)
@@ -237,6 +239,7 @@ func TestWriteRemove(t *testing.T) {
 	// create the multi writer
 	mwriter := MultiWriter(writers...)
 
+	// write and ensure io.Copy returns
 	write(t, mwriter, []byte(base))
 	wg.Wait()
 
@@ -249,6 +252,7 @@ func TestWriteRemove(t *testing.T) {
 		}
 	}
 
+	// write and ensure io.Copy returns
 	write(t, mwriter, []byte(dynamic))
 	wgRemoved.Wait()
 
@@ -285,8 +289,8 @@ func TestMultiRead(t *testing.T) {
 	mreader := MultiReader(readers...)
 
 	expected := strings.Repeat(base, count)
+	// read and ensure io.Copy returns
 	buffer := read(t, mreader, len(expected))
-
 	wg.Wait()
 
 	assert.Equal(t, expected, string(buffer))
@@ -330,6 +334,7 @@ func TestReadAdd(t *testing.T) {
 	mreader := MultiReader(readersLeft...)
 
 	expected := strings.Repeat(base, len(readersLeft))
+	// read and ensure io.Copy returns
 	buffer := read(t, mreader, len(expected))
 	wg.Wait()
 
@@ -344,8 +349,8 @@ func TestReadAdd(t *testing.T) {
 	}
 
 	expected = strings.Repeat(dynamic, len(readersAdded))
+	// read and ensure io.Copy returns
 	buffer = read(t, mreader, len(expected))
-
 	wgAdded.Wait()
 
 	assert.Equal(t, expected, string(buffer))
@@ -392,6 +397,7 @@ func TestReadRemove(t *testing.T) {
 	mreader := MultiReader(readers...)
 
 	expected := strings.Repeat(base, count)
+	// read and ensure io.Copy returns
 	buffer := read(t, mreader, len(expected))
 	wg.Wait()
 
@@ -405,8 +411,8 @@ func TestReadRemove(t *testing.T) {
 	}
 
 	expected = strings.Repeat(dynamic, len(readersLeft))
+	// read and ensure io.Copy returns
 	buffer = read(t, mreader, len(expected))
-
 	wgRemoved.Wait()
 
 	assert.Equal(t, expected, string(buffer))

--- a/pkg/dio/reader.go
+++ b/pkg/dio/reader.go
@@ -1,4 +1,4 @@
-// Copyright 2016 VMware, Inc. All Rights Reserved.
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/pkg/dio/reader.go
+++ b/pkg/dio/reader.go
@@ -55,7 +55,7 @@ func (t *multiReader) Read(p []byte) (int, error) {
 	t.mutex.Unlock()
 
 	// Close sets this
-	if err == io.EOF {
+	if err == io.EOF || err == io.ErrClosedPipe {
 		if verbose {
 			log.Debugf("[%p] read from closed multi-reader, returning EOF", t)
 		}
@@ -95,7 +95,7 @@ func (t *multiReader) Read(p []byte) (int, error) {
 		x, err := r.Read(slice)
 		n += x
 		if err != nil {
-			if err != io.EOF {
+			if err != io.EOF && err != io.ErrClosedPipe {
 				t.mutex.Lock()
 				// if there was an actual error, return that
 				t.err = err

